### ISSUE TITLE
How is the ingest job linked to a dataset

### ISSUE THREAD

**Issue by amanharshx** [ultralytics/ultralytics#24934]
I was playing around with the Platform REST API trying to populate a dataset programmatically. I created a dataset, uploaded an archive via the signed-URL flow, and then went to call the ingest endpoint — and got stuck on how ingest actually knows which dataset to put the data into.
  
**What the Ultralytics assistant told me:** the endpoint is `POST /api/datasets/{datasetId}/ingest` — with the dataset id in the URL path.
  
**What the docs actually say** (`docs/en/platform/api/index.md`, "Dataset Ingest"):
  
> `POST /api/datasets/ingest`
> The request body takes exactly one of `sessionId` or `sourceUrl`, plus an optional `targetSplit`.
  
So the docs show **no `datasetId`** in the path or the body, which contradicts the assistant's `/api/datasets/{datasetId}/ingest`. Because of that, it's unclear which dataset the ingested data actually ends up in.
  
### Questions
  
1. **Which endpoint is correct** — `POST /api/datasets/ingest` (per docs) or `POST /api/datasets/{datasetId}/ingest` (per the assistant)?
  
2. **`sessionId` case:** Is the target dataset inferred from the `assetId` passed earlier to `POST /api/upload/signed-url` (`assetType: "datasets"`, `assetId: <datasetId>`)? i.e. is the upload session
already bound to that dataset?
  
3. **`sourceUrl` case:** With a remote ZIP/TAR/NDJSON there's no upload session and no `datasetId`
anywhere in the request. Does ingest **create a new dataset automatically**, or import into an
**existing** one? If existing, how is the target dataset specified?
  
4. Could the docs add an explicit end-to-end example (create dataset → upload/ingest → poll status)
showing exactly where the `datasetId` linkage happens?

### Environment
- Platform REST API
- Reference: `docs/en/platform/api/index.md` → Dataset Ingest / Upload API sections

**Comment by UltralyticsAssistant** [comment]
👋 Hello @amanharshx, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue triaged quickly, and an Ultralytics engineer will also assist you soon.
It looks like this may be a documentation/API clarification request related to Platform dataset ingest behavior. We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
For this specific Platform API question, it would be especially helpful if you could share:
- The exact request sequence you used for dataset creation, signed-URL upload, and ingest
- The exact endpoint paths and request bodies used at each step
- Any response payloads or error messages returned by the API
- A minimal reproducible example showing the ambiguity between the documented ingest flow and the assistant-provided path
- Whether this was tested against the public Platform REST API docs reference in `docs/en/platform/api/index.md`
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the clear report @amanharshx. The docs were missing an important body field, and the assistant-provided path was not correct.

Ground truth for the current Platform API:

- Correct endpoint: `POST /api/datasets/ingest`
- Target dataset: pass `datasetId` in the JSON body, not in the URL path
- Uploaded archive flow: first call `POST /api/upload/signed-url` with `assetType: "datasets"` and `assetId: <datasetId>`. That upload session is bound to the dataset, and ingest still validates that the body `datasetId` matches the session `assetId`.
- Remote `sourceUrl` flow: ingest does not create a dataset automatically. Create the dataset first with `POST /api/datasets`, then call ingest with `{ "datasetId": "...", "sourceUrl": "..." }`.

So the body is either:

```json
{ "datasetId": "dataset_abc123", "sessionId": "session_abc123" }
```

or:

```json
{ "datasetId": "dataset_abc123", "sourceUrl": "https://example.com/my-dataset.zip" }
```

I opened #24943 to update the public docs with this flow and add the end-to-end linkage so this is explicit.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24943: Clarify Platform dataset ingest API flow](https://github.com/ultralytics/ultralytics/pull/24943) by @glenn-jocher.

This PR updates the Ultralytics Platform dataset ingest API documentation and may resolve the confusion behind this issue by clarifying the expected workflow and request structure, including:

- adding a documented response example for `POST /api/datasets`
- clarifying that `POST /api/datasets/ingest` is only for an existing dataset
- documenting that `datasetId` must be included in the ingest request body
- making it explicit that exactly one of `sessionId` or `sourceUrl` must be provided
- clarifying that `targetSplit` is optional
- separating the two ingest paths more clearly:
  - uploaded archive flow via `sessionId`
  - remote import flow via `sourceUrl`
- adding request examples for both upload-based and remote import workflows
- adding a response example for ingest with fields like `jobId`, `datasetId`, and `status`
- updating the workflow diagram to show the full sequence:
  - create dataset
  - request signed upload URL
  - upload file
  - complete upload
  - start ingest
- clarifying that `POST /api/upload/complete` only finalizes the upload session, and that `POST /api/datasets/ingest` must still be called to actually begin processing

Credit to @glenn-jocher for the fix and to everyone who helped report, investigate, and improve the docs around this flow.

If you'd like to test the latest changes now, you can install directly from `main` with:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Otherwise, it should also be available in the next official PyPI release.

Please let us know whether these changes resolve the issue on your side. Thanks 🙏 for reporting this, and any further feedback is welcome if the issue still persists.

**Comment by amanharshx** [comment]
@glenn-jocher Thanks for quick docs fix. Ingest flow is clear now 👍

**Comment by amanharshx** [comment]
@glenn-jocher 

Sorry to tack another question onto this thread (and thanks again — the archive ingest flow you described works perfectly: zipped images → signed-url(assetType:"datasets") → PUT → complete → ingest {datasetId, sessionId} → dataset goes to ready with the right imageCount).
  
While testing I hit one more gap, this time for **individual images** (not archives), and the assistant's guidance didn't pan out — so I wanted to check with you directly.
  
**What the assistant suggested (didn't work):**
  
- It first said to use POST /api/datasets/{datasetId}/upload with a multipart file — but that path isn't a real API route (it returns the web app HTML).
- It also claimed ingest takes a singular assetType (image/video/archive) — but the ingest body has no assetType, and it rejects an image upload session ("Upload session not found or expired").
  
**What I actually tried** (the documented signed-URL flow, since assetType: "images" is listed):
  
1. `POST /api/upload/signed-url` with 
`{ "assetType": "images", "assetId": "<datasetId>", "filename", "contentType", "totalBytes" }` → 
returns `sessionId + uploadUrl` ✅
2. `PUT` the image bytes to `uploadUrl` → `200` ✅
3. `POST /api/upload/complete` with `{ "sessionId" }` → 
`{ "success": true, "file": { "path": "gs://.../images/<datasetId>/img.webp", ... } }` ✅
  
The bytes land in cloud storage, but the **image never attaches to the dataset** — imageCount stays 0 and the dataset is empty in both the API and the UI, even after waiting. I also couldn't find a "register image" endpoint (POST /api/datasets/{id}/images seems to want some array body I couldn't figure out).
  
So my question: **is there a public REST flow to add individual images/videos to a dataset, or is that currently UI-only — with archive ingest being the intended programmatic path?** If it's UI-only for now that's totally fine; I'll just document the zip-and-ingest approach. Just want to be sure I'm not missing an endpoint. 

Thanks again for your patience!

**Comment by glenn-jocher** [comment]
Based on the current [REST API reference](https://docs.ultralytics.com/platform/api) and [Datasets docs](https://docs.ultralytics.com/platform/data/datasets), the public programmatic flow today is archive or NDJSON upload followed by `POST /api/datasets/ingest`; there is not currently a documented public REST endpoint to attach standalone image or video uploads to a dataset after `POST /api/upload/complete`, so from a public-API perspective that path is effectively UI-only for now, and please open a new issue if you’d like us to track documenting or exposing single-file upload support.
